rfcs/address_space_overrides.md

```diff
@@ -0,0 +1,111 @@
+it to artificially consider specific IP addresses `local`, `private` or
+`public`.
+
+This RFC aims to fix
+[web-platform-tests/wpt#26166](https://github.com/web-platform-tests/wpt/issues/26166).
+
+## Details
+
+### Address allocation
+
+For the purposes of testing, declare that the following subnets are not `local`
+but instead:
+
+* 127.1.0.0/16: `private`
+* 127.2.0.0/16: `public`
+
+### Browser changes
+
+A new command-line flag (or a collection thereof) is added to each browser (it
+need not be uniform, though that would certainly help with implementation) that
+allows overriding the address space derived from specific IP addresses. For
+example:
+
+```sh
+--test-override-address-space=127.1.0.0/16:private,127.2.0.0/16:public
+```
+
+Test-only code wires this command-line flag to the code in the browser under
+test that determines the address space of a network endpoint.
+
+### Test runner changes
+
+The web platform test runner sets command-line flags on browsers such that the
+browsers implement the custom address space mapping described above during
+tests.
+
+Corresponding entries are added to the hosts file, as generated by
+[make_hosts_file()](https://github.com/web-platform-tests/wpt/blob/master/tools/serve/serve.py#L513-L530):
+
+```
+127.1.0.1 private.web-platform.test
+127.2.0.1 public.web-platform.test
+```
+
+This allows web platform tests to exercise browser behavior in the face of
+`private` and `public` IP addresses, simply by targeting the above domains.
+
+## Risks
+
+If the tests run on an IPv6-only system, the system might not know how to route
+requests to the IPv4 loopback address range? This seems like an extremely remote
+possibility, in which case many other assumptions made by the test framework
+might have to be revisited anyway.
+
+## Alternatives considered
+
+### Use individual IP addresses
+
+Same as above, but allow overriding address spaces per IP address instead of
+per subnet.
+
+The marginal difficulty of overriding per subnet seems low, and provides plenty
+of test IP addresses to exercise intra-address-space, cross-ip-address requests,
+which might come in handy [in the
+future](https://github.com/WICG/cors-rfc1918/pull/1#issuecomment-721110250).
+
+### Use real IP addresses
+
+If we need to test user agent behavior in the face of a variety of IP addresses,
+then... just do it!
+
+This approach would have the web platform test runner configure additional
+loopback network devices and assign them `private` and `public` IP addresses.
+Then we could configure the hosts file to resolve specific domains to those IP
+addresses, and test the whole shebang end to end.
+
+This approach's main pro is that it truly is an end to end test of the feature,
+exercising production code to its full extent. By virtue of requiring no
+test-only hooks, it also requires no work on the part of browser developers.
+
+Its main con is that it requires the test runner to have system privileges.
```

Comment:
That is true, but is it so unacceptable? In the automated test execution environment, we for sure can do it. For developers needing to test this locally, we can have instructions/an automated script to do it (similarly as the make /etc/hosts script). And we can document that those tests will fail otherwise if run locally.

Reply:
AFAIK, LUCI Swarming bot doesn't give test code root permissions, but I'm not sure if that's a strong constraint. Also cc @jgraham on Gecko infra.

Reply:
I'm incredibly reluctant to have any requirement that the test harness runs with root permissions. Historically we've avoided this requirement even when there are tradeoffs in terms of testability (e.g. we don't run the server on ports 80/443, which limit our ability to test the different URL serializations in those cases). 

On gecko infra it *might* be possible to get root, although I'm pretty sure nothing does at the moment. We use a gecko pref for the DNS rather than editing the system `/etc/hosts`. For local gecko development there's no requirement to get root access except for installing the build tools. For web-platform-tests in particular you might need to add the hosts if you want to test in a non-Firefox browser or use the server standalone, but it isn't a requirement.

Requiring people to reconfigure their system network setup, or requiring the test runner to have root access so it can fiddle with the system setup, both seem like they have a high chance of ending up with user complaints either at the principle, or because the changes break their system in some way. The `/etc/hosts` modifications are borderline acceptable in that the actual modification is straightforward (just appending some lines to the hosts file) and works the same way on all systems we support. I'm extremely reluctant to add anything else which requires root, especially when it's not so easy to understand or debug.

Reply:
I was not proposing to have the test runner run as root, but rather to have this setup as a separate step, and to have the infrastructure already set up for this.

Regarding local setup, I imagine not every developer wants to run every test on his machine, so this would be a problem of a very small number of people. But I do understand there are some cons here.

Reply:
See my other comment for the downside of having this setup as a separate step (outside of test runner). The test runner still needs internet access, so we'd need some refactoring to pull that out; otherwise, we run the risk of obscuring legit IPs if we don't own them.

Reply:
I'll add some detail to this section to reflect this discussion.

Reply:
Done.